scala/lambdas/ingest-mapper/src/main/scala/uk/gov/nationalarchives/ingestmapper/MetadataService.scala

```diff
@@ -59,119 +59,114 @@ class MetadataService(s3: DAS3Client[IO], discoveryService: DiscoveryService[IO]
+        val depSeriesCount = idToDepartmentSeries.values.toList
+          .groupBy(_.departmentItem("id").str)
+          .view
+          .mapValues(_.distinct.length)
+          .toMap
+        val parentToChildCount = idToParent.groupBy(_._2).view.mapValues(_.size).toMap ++ depSeriesCount
+
+        def addChildCount(obj: Obj) = {
+          val attributes = obj.value.toMap
+          val id = attributes("id").str
+          val parentPath = attributes.get("parentPath").map(parentPath => s"${parentPath.str}/").getOrElse("")
+          val childParentPath = s"$parentPath$id"
+          val childCount = parentToChildCount.getOrElse(childParentPath, 1)
+          Obj.from(attributes + ("childCount" -> Num(childCount)))
         }
-    )
 
-  private def parseFileFromS3[T](input: Input, decoderPipe: Pipe[IO, String, T]): IO[List[T]] =
+        val departmentSeriesObjects = idToDepartmentSeries.values.toList
+          .distinctBy(item => item.potentialSeriesItem.map(obj => obj("name")))
+          .flatMap { departmentSeries =>
+            List(
+              departmentSeries.potentialSeriesItem.map(addChildCount),
+              Option(addChildCount(departmentSeries.departmentItem))
+            ).flatten
+          }
+          .distinct
+        jsonArr.map { metadataEntry =>
+          val id = UUID.fromString(metadataEntry("id").str)
+          val name = metadataEntry("name").str
+          val parentPath = idToParent(id)
+          val checksum: Value = Try(metadataEntry("checksum_sha256")).toOption
+            .map(_.str)
+            .map(Str.apply)
+            .getOrElse(Null)
+          val entryType = metadataEntry("type").str
+          val fileExtension =
+            if (entryType == File.toString)
+              name.split('.').toList.reverse match {
+                case ext :: _ :: _ => Str(ext)
+                case _ => Null
+              }
+            else Null
+          val childCount = Num(parentToChildCount.getOrElse(s"$parentPath/$id", 0))
+          val metadataMap: Map[String, Value] =
+            Map(
+              "batchId" -> Str(input.batchId),
+              "parentPath" -> Str(parentPath),
+              "checksum_sha256" -> checksum,
+              "fileExtension" -> fileExtension,
+              "childCount" -> childCount
+            ) ++ metadataEntry.obj.view
+              .filterKeys(_ != "parentId")
+              .toMap
+          Obj.from(metadataMap)
+        } ++ departmentSeriesObjects
+      }
+    }
+
+  }
+
+  private def metadataFromS3[T](input: Input): IO[String] =
     for {
       pub <- s3.download(input.metadataPackage.getHost, input.metadataPackage.getPath.drop(1))
       s3FileString <- pub
         .toStreamBuffered[IO](bufferSize)
         .flatMap(bf => Stream.chunk(Chunk.byteBuffer(bf)))
         .through(text.utf8.decode)
-        .through(decoderPipe)
         .compile
         .toList
-    } yield s3FileString
+    } yield s3FileString.mkString
```

Comment:
Just to confirm, is `s3FileString` a `List` of 5MB strings and then `mkString` is merging them into one string? If so, maybe we can call it `fileAsStrings` or `listOfFileStrings`?